Docs: Settings updates at runtime (#34169)
grafana/grafana: docs/sources/enterprise/settings-updates.md

+++
title = "Settings updates at runtime"
description = "Settings updates at runtime"
keywords = ["grafana", "runtime", "settings"]
weight = 500
+++

# Settings updates at runtime

> Only available in Grafana Enterprise v8.0+.

Settings updates at runtime allows you to update Grafana settings with no need to restart the instance.

Currently, **it only supports updates on the `auth.saml` section.** 

## Update settings via the API

You can update / remove settings through the [Settings API]({{< relref "../http_api/settings.md#update--remove-database-settings" >}}).

It verifies if the given settings updates (or removals) are allowed and valid, persists them into the database and reload
Grafana services with no need to restart the instance.

> 

Author: I added few more lines describing the API and the expected requests payload, so feel free to come back to me again with additional feedback. Thanks!! 🙌🏻

+++
title = "Settings updates at runtime"
description = "Settings updates at runtime"
keywords = ["grafana", "runtime", "settings"]
weight = 500
+++

# Settings updates at runtime

> **Note:** Available in Grafana Enterprise v8.0+.

> Settings updates at runtime allows you to update Grafana settings with no need to restart the instance.

Reviewer: I just realized that we should mention that the db settings will override settings from other sources and that if db settings are removed they config will revert to whatever is set in the underlying layers of config.

```db < commandline < env vars < file < defaults)```

Author: Added, good catch! Thanks! 👌🏻

Currently, **it only supports updates on the `auth.saml` section.** 

## Update settings via the API

You can update settings through the [Admin API]({{< relref "../http_api/admin.md#update-settings" >}}).

It verifies if the given settings updates are allowed and valid, persists them into the database and reloads
Grafana services with no need to restart the instance.

So, the payload of a `PUT` request to the update settings endpoint (`/api/admin/settings`) 
should contain (either one or both):
- An `updates` map with a key, and a value per section you want to set.

> - A `removals` list with keys per section you want to set.

Reviewer: ```suggestion
- A `removals` list with keys per section you want to unset.
```

+++
title = "Settings updates at runtime"
description = "Settings updates at runtime"
keywords = ["grafana", "runtime", "settings"]
weight = 500
+++

# Settings updates at runtime

> **Note:** Available in Grafana Enterprise v8.0+.

Settings updates at runtime allows you to update Grafana settings with no need to restart the instance.

Those updates that happen at runtime are stored in the database and override settings from the other sources

> (arguments, environment variables, settings file, etc). Therefore, every time a specific setting key is removed at runtime,

Reviewer 2: Would be nice to add links to these config methods, in case people don't know what they are.

Author: Yeah, great idea! Do you think a link to the general configuration page would be enough?
https://grafana.com/docs/grafana/latest/administration/configuration/

Or would be better to add a specific link to every alternative?
- https://grafana.com/docs/grafana/latest/administration/configuration/#config-file-locations
- https://grafana.com/docs/grafana/latest/administration/configuration/#configure-with-environment-variables
- https://grafana.com/docs/grafana/latest/administration/configuration/#variable-expansion
- https://grafana.com/docs/grafana/latest/administration/cli/#override-a-configuration-setting

I think adding the link/s to the first occurrence would be enough, right? 🤔

Reviewer 2: Yes, linking once to a single page that includes all of the config options makes sense.

+++
title = "Settings updates at runtime"
description = "Settings updates at runtime"
keywords = ["grafana", "runtime", "settings"]
weight = 500
+++

# Settings updates at runtime

> **Note:** Available in Grafana Enterprise v8.0+.

> Settings updates at runtime allows you to update Grafana settings with no need to restart the instance.

Reviewer 2: ```suggestion
Settings updates at runtime allows you to update Grafana settings with no need to restart the Grafana server.
```

> Those updates that happen at runtime are stored in the database and override

Reviewer 2: ```suggestion
Updates that happen at runtime are stored in the database and override
```

[settings from the other sources](https://grafana.com/docs/grafana/latest/administration/configuration/)
(arguments, environment variables, settings file, etc). Therefore, every time a specific setting key is removed at runtime,
the value used for that key is the inherited one from the other sources in the reverse order of precedence 
(`arguments > environment variables > settings file`), being the application default the value used when no one provided
through one of these, at least.

Currently, **it only supports updates on the `auth.saml` section.** 

## Update settings via the API

You can update settings through the [Admin API]({{< relref "../http_api/admin.md#update-settings" >}}).

> It verifies if the given settings updates are allowed and valid, persists them into the database and reloads

Reviewer 2: ```suggestion
When you submit a settings update via API, Grafana verifies if the given settings updates are allowed and valid. If they are, then Grafana stores the settings in the database and reloads
```

should contain (either one or both):
- An `updates` map with a key, and a value per section you want to set.
- A `removals` list with keys per section you want to unset.

For example, if you provide the following `updates`:

```json
{
  "auth.saml": {
    "enabled": "true",
    "single_logout": "false"
}
```
it would enable SAML and disable single logouts. And, if you provide the following `removals`:
```json
{
  "auth.saml": ["allow_idp_initiated"]
}
```

it would remove the key/value setting identified by `allow_idp_initiated` within the `auth.saml`.
So, the SAML service would be reloaded and that value would be inherited for either (settings `.ini` file,
environment variable, command line arguments or any other accepted mechanism to provide configuration).

Therefore, the complete HTTP payload would looks like:

```json
{
  "updates": {
    "auth.saml": {
      "enabled": "true",
      "single_logout": "false"
    }
  },
  "removals": {
    "auth.saml": ["allow_idp_initiated"]
  }
}
```

In case any of these settings cannot be overridden nor valid, it would return an error and these settings
won't be persisted into the database.

## Background job (high availability set-ups)

> Grafana Enterprise has a built-in scheduled background job that every minute looks into the database for

Reviewer 2: ```suggestion
Grafana Enterprise has a built-in scheduled background job that looks into the database every minute for
```

> settings updates. If so, it reloads the Grafana services affected by the detected changes. 

Reviewer 2: ```suggestion
settings updates. If there are updates, it reloads the Grafana services affected by the detected changes. 
```

> It is used as the synchronization mechanism in high availability set-ups. So, after you perform some changes through the

Reviewer 2: ```suggestion
The background job synchronizes settings between instances in high availability set-ups. So, after you perform some changes through the
```
